# Loading Modules

In [1]:
import os
import sys 

sys.path.append(os.path.abspath('../scripts'))

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from IPython.display import display
import warnings 

warnings.filterwarnings('ignore')


In [4]:
# Load the data

data_path = "data/Week2_challenge_data_source.xlsx"
data = pd.read_excel(data_path)

display(
    data.head(),
    data.info(),
    data.describe()
)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150001 entries, 0 to 150000
Data columns (total 55 columns):
 #   Column                                    Non-Null Count   Dtype         
---  ------                                    --------------   -----         
 0   Bearer Id                                 150001 non-null  object        
 1   Start                                     150000 non-null  datetime64[ns]
 2   Start ms                                  150000 non-null  float64       
 3   End                                       150000 non-null  datetime64[ns]
 4   End ms                                    150000 non-null  float64       
 5   Dur. (ms)                                 150000 non-null  float64       
 6   IMSI                                      149431 non-null  float64       
 7   MSISDN/Number                             148935 non-null  float64       
 8   IMEI                                      149429 non-null  float64       
 9   Last Location N

Bearer Id               Start  Start ms                 End  \
0  13114483460844900352 2019-04-04 12:01:18     770.0 2019-04-25 14:35:31   
1  13114483482878900224 2019-04-09 13:04:04     235.0 2019-04-25 08:15:48   
2  13114483484080500736 2019-04-09 17:42:11       1.0 2019-04-25 11:58:13   
3  13114483485442799616 2019-04-10 00:31:25     486.0 2019-04-25 07:36:35   
4  13114483499480700928 2019-04-12 20:10:23     565.0 2019-04-25 10:40:32   

   End ms  Dur. (ms)          IMSI  MSISDN/Number          IMEI  \
0   662.0  1823652.0  2.082014e+14   3.366496e+10  3.552121e+13   
1   606.0  1365104.0  2.082019e+14   3.368185e+10  3.579401e+13   
2   652.0  1361762.0  2.082003e+14   3.376063e+10  3.528151e+13   
3   171.0  1321509.0  2.082014e+14   3.375034e+10  3.535661e+13   
4   954.0  1089009.0  2.082014e+14   3.369980e+10  3.540701e+13   

  Last Location Name  ...  Youtube DL (Bytes)  Youtube UL (Bytes)  \
0   9164566995485190  ...          15854611.0           2501332.0   
1            L77566A  ...          20247395.0          19111729.0   
2            D42335A  ...          19725661.0          14699576.0   
3            T21824A  ...          21388122.0          15146643.0   
4            D88865A  ...          15259380.0          18962873.0   

   Netflix DL (Bytes)  Netflix UL (Bytes)  Gaming DL (Bytes)  \
0           8198936.0           9656251.0        278082303.0   
1          18338413.0          17227132.0        608750074.0   
2          17587794.0           6163408.0        229584621.0   
3          13994646.0           1097942.0        799538153.0   
4          17124581.0            415218.0        527707248.0   

   Gaming UL (Bytes)  Other DL (Bytes)  Other UL (Bytes)  Total UL (Bytes)  \
0         14344150.0       171744450.0         8814393.0        36749741.0   
1          1170709.0       526904238.0        15055145.0        53800391.0   
2           395630.0       410692588.0         4215763.0        27883638.0   
3         10849722.0       749039933.0        12797283.0        43324218.0   
4          3529801.0       550709500.0        13910322.0        38542814.0   

   Total DL (Bytes)  
0       308879636.0  
1       653384965.0  
2       279807335.0  
3       846028530.0  
4       569138589.0  

[5 rows x 55 columns]

None

Start       Start ms  \
count                         150000  150000.000000   
mean   2019-04-26 12:36:38.953740032     499.188200   
min              2019-04-04 12:01:18       0.000000   
25%    2019-04-25 00:26:00.750000128     250.000000   
50%       2019-04-26 08:51:24.500000     499.000000   
75%    2019-04-27 18:18:59.750000128     749.000000   
max              2019-04-29 07:28:43     999.000000   
std                              NaN     288.611834   

                                 End         End ms     Dur. (ms)  \
count                         150000  150000.000000  1.500000e+05   
mean   2019-04-27 17:40:08.059673344     498.800880  1.046086e+05   
min              2019-04-24 22:59:58       0.000000  7.142000e+03   
25%       2019-04-26 07:30:11.500000     251.000000  5.744050e+04   
50%              2019-04-27 23:02:05     500.000000  8.639900e+04   
75%       2019-04-29 07:15:07.500000     750.000000  1.324302e+05   
max              2019-04-30 23:59:27     999.000000  1.859336e+06   
std                              NaN     288.097653  8.103762e+04   

               IMSI  MSISDN/Number          IMEI  Avg RTT DL (ms)  \
count  1.494310e+05   1.489350e+05  1.494290e+05    122172.000000   
mean   2.082016e+14   4.188282e+10  4.847455e+13       109.795706   
min    2.040471e+14   3.360100e+10  4.400152e+11         0.000000   
25%    2.082014e+14   3.365130e+10  3.546071e+13        32.000000   
50%    2.082015e+14   3.366371e+10  3.572201e+13        45.000000   
75%    2.082018e+14   3.368349e+10  8.611970e+13        70.000000   
max    2.140743e+14   8.823971e+14  9.900120e+13     96923.000000   
std    2.148809e+10   2.447443e+12  2.241637e+13       619.782739   

       Avg RTT UL (ms)  ...  Youtube DL (Bytes)  Youtube UL (Bytes)  \
count    122189.000000  ...        1.500010e+05        1.500010e+05   
mean         17.662883  ...        1.163407e+07        1.100941e+07   
min           0.000000  ...        5.300000e+01        1.050000e+02   
25%           2.000000  ...        5.833501e+06        5.517965e+06   
50%           5.000000  ...        1.161602e+07        1.101345e+07   
75%          15.000000  ...        1.744852e+07        1.651556e+07   
max        7120.000000  ...        2.325910e+07        2.201196e+07   
std          84.793524  ...        6.710569e+06        6.345423e+06   

       Netflix DL (Bytes)  Netflix UL (Bytes)  Gaming DL (Bytes)  \
count        1.500010e+05        1.500010e+05       1.500010e+05   
mean         1.162685e+07        1.100175e+07       4.220447e+08   
min          4.200000e+01        3.500000e+01       2.516000e+03   
25%          5.777156e+06        5.475981e+06       2.104733e+08   
50%          1.164222e+07        1.099638e+07       4.234081e+08   
75%          1.747048e+07        1.650727e+07       6.331742e+08   
max          2.325919e+07        2.201196e+07       8.434419e+08   
std          6.725218e+06        6.359490e+06       2.439675e+08   

       Gaming UL (Bytes)  Other DL (Bytes)  Other UL (Bytes)  \
count       1.500010e+05      1.500010e+05      1.500010e+05   
mean        8.288398e+06      4.211005e+08      8.264799e+06   
min         5.900000e+01      3.290000e+03      1.480000e+02   
25%         4.128476e+06      2.101869e+08      4.145943e+06   
50%         8.291208e+06      4.218030e+08      8.267071e+06   
75%         1.243162e+07      6.316918e+08      1.238415e+07   
max         1.655879e+07      8.434425e+08      1.655882e+07   
std         4.782700e+06      2.432050e+08      4.769004e+06   

       Total UL (Bytes)  Total DL (Bytes)  
count      1.500000e+05      1.500000e+05  
mean       4.112121e+07      4.546434e+08  
min        2.866892e+06      7.114041e+06  
25%        3.322201e+07      2.431068e+08  
50%        4.114331e+07      4.558411e+08  
75%        4.903424e+07      6.657055e+08  
max        7.833131e+07      9.029696e+08  
std        1.127639e+07      2.441429e+08  

[8 rows x 51 columns]

### Observations:
1. **Number of Rows and Columns**:
   - The dataset has 150,001 rows and 55 columns.

2. **Missing Data**:
   - Several columns have missing values. For instance:
     - Columns like `Avg RTT DL (ms)` and `Avg RTT UL (ms)` have ~122k non-null entries out of 150k.
     - `TCP DL Retrans. Vol (Bytes)` and `TCP UL Retrans. Vol (Bytes)` have even fewer entries.
   - Columns like `Social Media DL (Bytes)` and `Total DL (Bytes)` have complete data.

3. **Data Types**:
   - Most columns are numeric (`float64`), with a few categorical (`object`) and datetime (`datetime64[ns]`) columns.

4. **Potential Redundancies**:
   - Some columns like `Dur. (ms)` and `Dur. (ms).1` appear to be duplicated or redundant.

5. **Key Variables**:
   - Usage metrics such as `Avg Bearer TP DL (kbps)`, `Avg Bearer TP UL (kbps)`, and `Total DL/UL (Bytes)` are critical for identifying user behavior.
   - Columns such as `Handset Manufacturer` and `Handset Type` can provide insights into device preferences.

6. **Customer Segmentation Opportunities**:
   - Metrics like `Social Media DL (Bytes)` or `Gaming DL (Bytes)` could help identify customer usage patterns.


# Task 1: User Overview Analysis 



## Task 1.1 - 

Your employer wants to have an overview of the users’ behavior on those applications.   
Aggregate per user the following information in the column  
- number of xDR sessions
- Session duration
- the total download (DL) and upload (UL) data
- the total data volume (in Bytes) during this session for each application

In [20]:
def aggregate_user_metrics(data, group_col, metrics):
    """
    Generalized function to aggregate user metrics.

    Parameters:
        data (pd.DataFrame): The input dataset.
        group_col (str): The column to group by (e.g., 'MSISDN/Number').
        metrics (dict): Dictionary where keys are the new column names 
                        and values are the columns to aggregate.
                        For sums, use a single column; for totals, pass a tuple.

    Returns:
        pd.DataFrame: Aggregated metrics per user.
    """
    agg_dict = {}
    agg_dict[group_col] = data[group_col].nunique()
    for metric_name, cols in metrics.items():
        if metric_name == "xDR_sessions":
            # Count unique session identifiers
            agg_dict[metric_name] = data.groupby(group_col)[cols].nunique()
        elif isinstance(cols, tuple):
            dl_col, ul_col = cols
            agg_dict[metric_name] = data.groupby(group_col)[[dl_col, ul_col]].sum().sum(axis=1)
        else:
            agg_dict[metric_name] = data.groupby(group_col)[cols].sum()

    # Combine all metrics into a DataFrame
    aggregated_data = pd.DataFrame(agg_dict)
    aggregated_data[group_col] = aggregated_data.index

    return aggregated_data.reset_index(drop=True)


# Updated metrics dictionary
metrics_to_aggregate = {
    'xDR_sessions': 'Bearer Id',  # Count unique sessions
    'Total_Session_Duration': 'Dur. (ms)',  # Sum session durations
    'Total_Data': ('Total DL (Bytes)', 'Total UL (Bytes)'),  # Sum total data volume
    'Social_Media_Data': ('Social Media DL (Bytes)', 'Social Media UL (Bytes)'),
    'Google_Data': ('Google DL (Bytes)', 'Google UL (Bytes)'),
    'Email_Data': ('Email DL (Bytes)', 'Email UL (Bytes)'),
    'YouTube_Data': ('Youtube DL (Bytes)', 'Youtube UL (Bytes)'),
    'Netflix_Data': ('Netflix DL (Bytes)', 'Netflix UL (Bytes)'),
    'Gaming_Data': ('Gaming DL (Bytes)', 'Gaming UL (Bytes)'),
    'Other_Data': ('Other DL (Bytes)', 'Other UL (Bytes)')
}

# Call the function
group_col = 'MSISDN/Number'

user_metrics = aggregate_user_metrics(data, group_col, metrics_to_aggregate)

display(
    user_metrics.shape,
    user_metrics.sample(20),
    user_metrics.describe()
)


(106856, 11)

MSISDN/Number  xDR_sessions  Total_Session_Duration    Total_Data  \
19952    3.364233e+10             1                 63897.0  9.082347e+08   
67341    3.366788e+10             1                135598.0  8.323755e+08   
84144    3.369803e+10             1                137019.0  4.550424e+08   
76283    3.367570e+10             2                291018.0  9.563160e+08   
98827    3.376290e+10             2                147318.0  8.532591e+08   
3782     3.361072e+10             1                 86399.0  6.735480e+07   
80224    3.368363e+10             1                 86399.0  5.703429e+08   
17114    3.363326e+10             1                142736.0  8.441520e+08   
44623    3.366154e+10             1                 86121.0  8.981107e+08   
11070    3.362236e+10             2                282748.0  4.502416e+08   
76806    3.367678e+10             1                108080.0  6.518957e+08   
67752    3.366801e+10             1                 86399.0  3.806260e+08   
90975    3.375307e+10             2                 66821.0  9.339370e+08   
12931    3.362541e+10             2                 53678.0  1.585258e+09   
62129    3.366615e+10             1                168496.0  6.875806e+08   
96713    3.376194e+10             1                172798.0  4.230224e+08   
44123    3.366143e+10             2                213607.0  1.287604e+09   
47708    3.366225e+10             1                 97472.0  9.036936e+08   
32906    3.365883e+10             4                314275.0  1.910984e+09   
105697   3.378393e+10             1                 22221.0  8.268549e+08   

        Social_Media_Data  Google_Data  Email_Data  YouTube_Data  \
19952           2145313.0   11366316.0    395819.0    14950695.0   
67341            879492.0    5777535.0   3260199.0    27162254.0   
84144           1203427.0    2245384.0    939393.0     3326220.0   
76283           2937284.0   18815659.0   4150020.0    42541573.0   
98827           7674565.0   19558868.0  10051749.0    53619543.0   
3782            2074668.0   10195704.0   1921514.0     9679016.0   
80224           1031421.0    6870278.0   1796520.0    20622995.0   
17114            312247.0    9916507.0   3418543.0    28142086.0   
44623            488603.0    3874553.0   3688238.0    27288956.0   
11070           4214750.0   14836590.0    845944.0    37683158.0   
76806           1887344.0    9879790.0   1379279.0    21898433.0   
67752           2555314.0   12195692.0   2899319.0    36714728.0   
90975           3233363.0   23023841.0   1545956.0    61090962.0   
12931           2668513.0    5859438.0   3384791.0    48405996.0   
62129           2657115.0    2731980.0   1160549.0    32795347.0   
96713           4559259.0   13349052.0   4699541.0    45197886.0   
44123           3414747.0    9567830.0   7041015.0    38440518.0   
47708           2010923.0    2866931.0   2661063.0     7788026.0   
32906           7999272.0   37672635.0   6932477.0    88209809.0   
105697          2753516.0    7697234.0   1290858.0    35421459.0   

        Netflix_Data   Gaming_Data    Other_Data  
19952     17029239.0  8.465179e+08  7.653637e+08  
67341     29956086.0  7.545764e+08  3.378061e+08  
84144     19048507.0  4.187879e+08  1.948199e+08  
76283     51551155.0  8.142703e+08  5.362436e+08  
98827     60921837.0  6.753697e+08  1.115840e+09  
3782      16389769.0  2.547218e+07  6.366510e+08  
80224     25905743.0  5.061768e+08  3.555653e+08  
17114     20119762.0  7.800336e+08  5.797682e+08  
44623     11852471.0  8.389819e+08  4.240170e+08  
11070     37731363.0  3.385389e+08  8.509694e+08  
76806      7156300.0  6.079963e+08  6.958580e+08  
67752     24216267.0  2.917382e+08  5.332628e+08  
90975     53915277.0  7.659313e+08  6.338358e+08  
12931     46051461.0  1.450944e+09  9.310430e+08  
62129     18072073.0  6.244441e+08  7.658760e+08  
96713     70317457.0  2.737748e+08  1.471848e+09  
44123     40524913.0  1.180368e+09  7.961260e+08  
47708     33364506.0  8.455223e+08  4.16907

MSISDN/Number   xDR_sessions  Total_Session_Duration    Total_Data  \
count   1.068560e+05  106856.000000            1.068560e+05  1.068560e+05   
mean    4.511474e+10       1.379951            1.461672e+05  6.909621e+08   
std     2.889423e+12       0.776847            1.863587e+05  4.910559e+08   
min     3.360100e+10       1.000000            7.142000e+03  3.324901e+07   
25%     3.365088e+10       1.000000            7.130800e+04  3.585499e+08   
50%     3.366365e+10       1.000000            1.027400e+05  6.179231e+08   
75%     3.368344e+10       2.000000            1.727990e+05  8.574351e+08   
max     8.823971e+14      18.000000            1.855375e+07  8.846226e+09   

       Social_Media_Data   Google_Data    Email_Data  YouTube_Data  \
count       1.068560e+05  1.068560e+05  1.068560e+05  1.068560e+05   
mean        2.547967e+06  1.088243e+07  3.148796e+06  3.155840e+07   
std         1.908038e+06  7.544861e+06  2.222400e+06  2.129492e+07   
min         1.563000e+03  4.033000e+04  1.817600e+04  7.890300e+04   
25%         1.211282e+06  5.942636e+06  1.674481e+06  1.863109e+07   
50%         2.303756e+06  9.586153e+06  2.799824e+06  2.680038e+07   
75%         3.307509e+06  1.321474e+07  3.847198e+06  3.792798e+07   
max         4.337478e+07  1.521919e+08  4.241878e+07  4.529588e+08   

       Netflix_Data   Gaming_Data    Other_Data  
count  1.068560e+05  1.068560e+05  1.068560e+05  
mean   3.153833e+07  5.997690e+08  5.985088e+08  
std    2.128956e+07  4.491505e+08  4.489221e+08  
min    1.845690e+05  3.063580e+05  1.490450e+05  
25%    1.855597e+07  2.880631e+08  2.855599e+08  
50%    2.671889e+07  5.423492e+08  5.398323e+08  
75%    3.797696e+07  7.773041e+08  7.753975e+08  
max    3.995191e+08  7.749432e+09  8.167878e+09

### Observations:
1. **Number of Users**:
   - The aggregated data shows 106,856 users (`MSISDN/Number`), which represents unique identifiers.

2. **xDR Sessions**:
   - The average user has approximately 1.38 sessions, with a maximum of 18 sessions.

3. **Session Duration**:
   - The mean session duration is approximately 146,167 ms (or about 146 seconds).
   - There's significant variance, as the maximum session duration is ~18.5 million ms (about 5 hours).

4. **Total Data Volume**:
   - Users download/upload an average of ~691 MB per session, with the heaviest users consuming ~8.8 GB.
   - Applications like `Netflix`, `Gaming`, and `YouTube` are the largest data consumers on average.

5. **Application-Specific Data**:
   - Application data volumes (`Netflix_Data`, `Gaming_Data`, etc.) align with usage trends in telecommunications:
     - `Netflix` and `Gaming` are high-data-consuming applications.
     - `Google` and `Email` have lower data consumption compared to multimedia applications.

### Interpretation:
1. **Heavy Users**:
   - Some users have an exceptionally high number of sessions or consume significant amounts of data, which may represent power users or users with premium plans.
   - Example: A user with `18 sessions` and over `8 GB` of data usage.

2. **Application Usage Patterns**:
   - High data consumption for `Netflix` and `Gaming` suggests these services are prime candidates for targeted marketing or premium packages.
   - Moderate usage of `Social Media` and `Google` indicates widespread, consistent usage across users.

3. **Session Counts and Engagement**:
   - Users with a higher session count could represent more engaged customers, potentially indicative of customer loyalty or higher service reliance.


## Task 1.2 

Conduct an exploratory data analysis on those data & communicate useful insights. Ensure that you identify and treat all missing values and outliers in the dataset by replacing them with the mean or any possible solution of the corresponding column.